File: testdata/clean/safe_app.py
import yaml
import hashlib


# Safe: parameterized query — the %s is a DB placeholder, not string formatting.
def get_user(cursor, name):
    cursor.execute("SELECT * FROM users WHERE name = %s", (name,))


# Safe: authorization CHECK, not a forced admin flag.
def check_permissions(user):
    is_admin = False
    if user.role == "admin":
        return True
    return is_admin


# Safe: empty-string guard, not a hardcoded credential.
def login(request):
    if request.password == "":
        return False
    credentials = ""
    return credentials != ""


# Safe: yaml.load with an explicit safe Loader.
def load_config(raw):
    return yaml.load(raw, Loader=yaml.SafeLoader)


# Safe: strong hash for non-credential data.
def fingerprint(data):
    return hashlib.sha256(data).hexdigest()

File: testdata/clean/safe_app.js
const crypto = require("crypto");

// Safe: authorization CHECK and a default-false flag — not forced admin.
function checkAdmin(user) {
  const isAdmin = false;
  if (user.role === "admin") {
    return true;
  }
  return isAdmin;
}

// Safe: empty-password guard, no hardcoded credential.
function authenticate(req) {
  if (req.body.password === "") {
    return false;
  }
  const credentials = "";
  return credentials !== "";
}

// Safe: JSON is a safe serialization format (not insecure deserialization).
function processData(req) {
  return JSON.parse(req.body);
}

// Safe: strong hash for non-credential data.
function fingerprint(data) {
  return crypto.createHash("sha256").update(data).digest("hex");
}

// Safe: API key sourced from the environment, not hardcoded.
const apiKey = process.env.API_KEY;

module.exports = { checkAdmin, authenticate, processData, fingerprint, apiKey };
